\name{calcExp}
\alias{calcExp}
\alias{relIsoformExp}
%- Also NEED an '\alias' for EACH other topic documented here.
\title{
  %%  ~~function to do ... ~~
  Estimate expression of gene splicing variants. 
}
\description{
  %%  ~~ A concise (1-5 lines) description of what the function does. ~~
  Estimate expression of gene splicing variants,
  assuming that the set of variants is known.
  \code{calcEx} estimates absolute expression and \code{relIsoformExp}
  the relative expression (i.e. the proportion of transcripts coming
  from each variant within a given gene).
}
\usage{
calcExp(distrs, genomeDB, pc, readLength)

relIsoformExp(eset, geneid='entrezid')
}
%- maybe also 'usage' for other objects documented here.
\arguments{
\item{distrs}{ List of fragment distributions as generated by the \code{getDistrs} function}
\item{genomeDB}{ \code{knownGenome} object containing annotated genome,
  as returned by the \code{procGenome} function. }
\item{pc}{Named vector of exon path counts}
\item{readLength}{ Read length in bp, e.g. in a paired-end experiment where
  75bp are sequenced on each end one would set \code{readLength=75}.}
\item{eset}{\code{ExpressionSet} containing absolute expression estimates, as returned by \code{calcExp}.}
\item{geneid}{Name of the column in \code{fData(eset)} containing the gene identifier.}
}
\value{
  Expression set with expression estimates.
  \code{featureNames} identify each transcript via
  RefSeq ids, and the \code{featureData} contains Entrez identifiers
  and the length of each transcript (in bp).
}
\references{
Rossell D. QASPER: Quantifying Alternative Splicing from Paired End
Reads. Technical report, IRB Barcelona (2010)
}
\author{
  Camille Stephan-Otto Attolini, Manuel Kroiss, David Rossell
}
\examples{
##---- Should be DIRECTLY executable !! ----
##-- ==>  Define data, use random,
##--	or do  help(data=index)  for the standard data sets.
}
% Add one or more standard keywords, see file 'KEYWORDS' in the
% R documentation directory.
\keyword{ models }% __ONLY ONE__ keyword per line
