\name{procBam}
\alias{procBam}
%- Also NEED an '\alias' for EACH other topic documented here.
\title{
  %%  ~~function to do ... ~~
  Process BAM object
}
\description{
  %%  ~~ A concise (1-5 lines) description of what the function does. ~~
  Process paired-end data stored in BAM object generated by scanBam. 
  Outputs RangedData object.
  }
\usage{
procBam(bam)
}
%- maybe also 'usage' for other objects documented here.
\arguments{
  \item{bam}{
    %%     ~~Describe \code{bamFileName} here~~
    BAM object generated by scanBam
}
}
\details{
%%  ~~ If necessary, more details than the description above ~~
}
\value{
%%  ~Describe the value returned
%%  If it is a LIST, use
%%  \item{comp1 }{Description of 'comp1'}
%%  \item{comp2 }{Description of 'comp2'}
%% ...
\item{reads}{A RangedData object with reads from fragments with both
  ends correctly aligned after splitting them
  by the corresponding CIGAR. Unique identifiers by fragment are
  stored.}
\item{frags}{A RangedData objects with start and end of fragments with
  both reads aligned.}
}
\references{
%% ~put references to the literature/web site here ~
}
\author{
  %%  ~~who you are~~
  Camille Stephan-Otto Attolini
}
\note{
%%  ~~further notes~~
}

%% ~Make other sections like Warning with \section{Warning }{....} ~

\seealso{
%% ~~objects to See Also as \code{\link{help}}, ~~~
}
\examples{
##---- Should be DIRECTLY executable !! ----
##-- ==>  Define data, use random,
##--	or do  help(data=index)  for the standard data sets.


}
% Add one or more standard keywords, see file 'KEYWORDS' in the
% R documentation directory.
\keyword{ ~paired-end sequencing }
\keyword{ ~SAM/BAM }% __ONLY ONE__ keyword per line
